drivers/adc/ad7616/iio_ad7616.c

```diff
@@ -0,0 +1,327 @@
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	return ad7616_write(dev, (uint8_t)reg, (uint16_t)val);
+}
+
+void iio_ad7616_event_handler(struct iio_desc *desc)
+{
+	iio_step(desc);
+}
+
+static int get_raw(void *device,
+		   char *buf,
+		   uint32_t len,
+		   const struct iio_ch_info *channel,
+		   intptr_t id)
+{
+	int ret;
+	uint8_t index = channel->ch_num < 8 ? channel->ch_num : channel->ch_num - 8;
+	struct ad7616_conversion_result res[8];
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	// TODO: we call this on every call, fetching all 16 inputs. Doesn't look smart
+	ret = ad7616_read_data_serial(dev, res, 1);
+	if (ret)
+		return ret;
+
+	if (channel->ch_num < 8)
+		return sprintf(buf, "%i", res[index].channel_a);
+
+	return sprintf(buf, "%i", res[index].channel_b);
+}
+
+static int get_scale(void *device,
+		     char *buf,
+		     uint32_t len,
+		     const struct iio_ch_info *channel,
+		     intptr_t id)
+{
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	enum ad7616_ch ch = (channel->ch_num < 8) ? channel->ch_num : AD7616_VB0 +
+			    channel->ch_num - 8;
+	enum ad7616_range range = (ch < AD7616_VB0) ? dev->va[ch] : dev->vb[ch -
+				  AD7616_VB0];
+
+	switch (range) {
+	case AD7616_10V:
+		return sprintf(buf, "10");
+	case AD7616_5V:
+		return sprintf(buf, "5");
+	case AD7616_2V5:
+		return sprintf(buf, "2.5");
+	default:
+		return -EINVAL;
+	}
+}
+
+static int set_scale(void *device,
+		     char *buf,
+		     uint32_t len,
+		     const struct iio_ch_info *channel,
+		     intptr_t id)
+{
+	int ret;
+	int val;
+	enum ad7616_range range;
+	enum ad7616_ch ch = (channel->ch_num < 8) ? channel->ch_num : AD7616_VB0 +
+			    channel->ch_num - 8;
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	ret = sscanf(buf, "%u", &val);
+	if (ret != 1)
+		return -EINVAL;
+
+	switch (val) {
+	case 10:
+		range = AD7616_10V;
+		break;
+	case 5:
+		range = AD7616_5V;
+		break;
+	case 2:
+		range = AD7616_2V5;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ad7616_set_range(dev, ch, range);
+}
+
+/**
+ * @brief	Read buffer data corresponding to AD7616 IIO device
+ * @param	iio_dev_data[in] - Pointer to IIO device data structure
+ * @return	0 in case of success, negative error code otherwise
+ */
+static int32_t iio_ad7616_submit_buffer(struct iio_device_data *iio_dev_data)
+{
+	uint16_t ret;
+	struct ad7616_conversion_result results[8];
+	uint16_t data[2];
+	uint32_t i = 0;
+	struct ad7616_iio_dev *iio_dev = iio_dev_data->dev;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	for (; i < iio_dev_data->buffer->samples; i++) {
+		ret = ad7616_read_data_serial(dev, results, 1);
+		if (ret)
+			return ret;
+
+		data[0] = results[0].channel_a;
+		data[1] = results[0].channel_b;
+
+		ret = iio_buffer_push_scan(iio_dev_data->buffer, data);
+		if (ret)
+			return ret;
+	}
```

Comment:
hmmm... so when working with buffers there is the possibility to receive request only for one channel, or two channels, or any combination of channels based on the scan mask. you will have to handle that scenario based on the scan mask and only push the data for the channels which are requested. the scan mask should be available in iio_dev_data structure. also, the data should be added in the buffer based on the channel's scan type.

Reply:
> the data should be added in the buffer based on the channel's scan type.

Not sure to understand this

Reply:
in this case you are adding the data for all the channels in the buffer.
however, the data in the buffer should be added based on the channel scan_mask which is available in iio_dev_data structure.
please see https://wiki.analog.com/resources/tools-software/linux-software/libiio_internals for more information about LibIIO buffers.
you can see as an example here: https://github.com/analogdevicesinc/no-OS/blob/master/drivers/accel/adxl355/iio_adxl355.c#L1019 where the data of a channel is added in the buffer based on the scan mask

Reply:
Ah yes ok, thanks